// @aragon/toolkit/dist/helpers/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ARAPP_FILE = exports.SOLIDITY_FILE = exports.ARTIFACT_FILE = exports.MANIFEST_FILE = exports.APM_INITIAL_VERSIONS = exports.EXECUTE_FUNCTION_NAME = exports.LAST_BLOCK_GAS_LIMIT_FACTOR = exports.DEFAULT_GAS_FUZZ_FACTOR = exports.REGISTRAR_GAS_LIMIT = exports.DEFAULT_IPFS_TIMEOUT = exports.DEFAULT_APM_REGISTRY = exports.ARAGON_DOMAIN = exports.NO_MANAGER = exports.ANY_ENTITY = exports.ZERO_ADDRESS = void 0;
const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.ZERO_ADDRESS = ZERO_ADDRESS;
const ANY_ENTITY = '0xffffffffffffffffffffffffffffffffffffffff';
exports.ANY_ENTITY = ANY_ENTITY;
const NO_MANAGER = '0x0000000000000000000000000000000000000000';
exports.NO_MANAGER = NO_MANAGER;
const ARAGON_DOMAIN = 'aragonid.eth';
exports.ARAGON_DOMAIN = ARAGON_DOMAIN;
const DEFAULT_APM_REGISTRY = 'aragonpm.eth';
exports.DEFAULT_APM_REGISTRY = DEFAULT_APM_REGISTRY;
const DEFAULT_IPFS_TIMEOUT = 300000;
exports.DEFAULT_IPFS_TIMEOUT = DEFAULT_IPFS_TIMEOUT;
const REGISTRAR_GAS_LIMIT = '1000000';
exports.REGISTRAR_GAS_LIMIT = REGISTRAR_GAS_LIMIT;
const DEFAULT_GAS_FUZZ_FACTOR = 1.5;
exports.DEFAULT_GAS_FUZZ_FACTOR = DEFAULT_GAS_FUZZ_FACTOR;
const LAST_BLOCK_GAS_LIMIT_FACTOR = 0.95;
exports.LAST_BLOCK_GAS_LIMIT_FACTOR = LAST_BLOCK_GAS_LIMIT_FACTOR;
const EXECUTE_FUNCTION_NAME = 'execute';
exports.EXECUTE_FUNCTION_NAME = EXECUTE_FUNCTION_NAME;
const APM_INITIAL_VERSIONS = ['0.0.1', '0.1.0', '1.0.0'];
exports.APM_INITIAL_VERSIONS = APM_INITIAL_VERSIONS;
const MANIFEST_FILE = 'manifest.json';
exports.MANIFEST_FILE = MANIFEST_FILE;
const ARTIFACT_FILE = 'artifact.json';
exports.ARTIFACT_FILE = ARTIFACT_FILE;
const SOLIDITY_FILE = 'code.sol';
exports.SOLIDITY_FILE = SOLIDITY_FILE;
const ARAPP_FILE = 'arapp.json';
exports.ARAPP_FILE = ARAPP_FILE;
