// @aragon/toolkit/dist/ipfs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GO_IMPL_DIST_URL = exports.GO_IMPL_DIST_VERSION = exports.NO_INSTALLATION_MSG = exports.DEFAULT_DAEMON_ARGS = exports.DAEMON_READY_OUTPUT = exports.DAEMON_START_TIMEOUT = exports.DEFAULT_GATEWAYS = exports.GATEWAY_FETCH_TIMEOUT_MSG = exports.GATEWAY_FETCH_TIMEOUT = void 0;
const GATEWAY_FETCH_TIMEOUT = 20000; // 20s

exports.GATEWAY_FETCH_TIMEOUT = GATEWAY_FETCH_TIMEOUT;
const GATEWAY_FETCH_TIMEOUT_MSG = 'Request timed out';
exports.GATEWAY_FETCH_TIMEOUT_MSG = GATEWAY_FETCH_TIMEOUT_MSG;
const DEFAULT_GATEWAYS = ['https://ipfs.io/ipfs', 'https://ipfs.infura.io/ipfs', 'https://cloudflare-ipfs.com/ipfs', 'https://ipfs.eth.aragon.network/ipfs', // 'https://ipfs.jes.xxx/ipfs',
'https://www.eternum.io/ipfs' // 'https://ipfs.wa.hle.rs/ipfs',
];
exports.DEFAULT_GATEWAYS = DEFAULT_GATEWAYS;
const DAEMON_START_TIMEOUT = 20000; // 20s

exports.DAEMON_START_TIMEOUT = DAEMON_START_TIMEOUT;
const DAEMON_READY_OUTPUT = 'Daemon is ready';
exports.DAEMON_READY_OUTPUT = DAEMON_READY_OUTPUT;
const DEFAULT_DAEMON_ARGS = ['--migrate'];
exports.DEFAULT_DAEMON_ARGS = DEFAULT_DAEMON_ARGS;
const NO_INSTALLATION_MSG = 'IPFS is not installed. Use `aragon ipfs install` before proceeding.';
exports.NO_INSTALLATION_MSG = NO_INSTALLATION_MSG;
const GO_IMPL_DIST_VERSION = '0.4.22';
exports.GO_IMPL_DIST_VERSION = GO_IMPL_DIST_VERSION;
const GO_IMPL_DIST_URL = 'https://dist.ipfs.io';
exports.GO_IMPL_DIST_URL = GO_IMPL_DIST_URL;
