// @aragon/toolkit/dist/node/packages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPackageRoot = exports.getNodePackageManager = exports.getGlobalPackagesLocation = exports.initPackage = exports.isPackage = void 0;

var _path = require("path");

var _execa = _interopRequireDefault(require("execa"));

var _fsExtra = require("fs-extra");

var _pkgDir = _interopRequireDefault(require("pkg-dir"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const isPackage = dir => (0, _path.join)(dir, 'package.json');

exports.isPackage = isPackage;

const initPackage = async path => {
  await (0, _fsExtra.ensureDir)(path);
  const {
    stdout
  } = await (0, _execa.default)('npm', ['init', '--yes'], {
    cwd: path
  });
  return stdout;
};

exports.initPackage = initPackage;

const getGlobalPackagesLocation = async () => {
  const {
    stdout
  } = await (0, _execa.default)('npm', ['prefix', '--global']);
  return stdout;
};

exports.getGlobalPackagesLocation = getGlobalPackagesLocation;

const getNodePackageManager = () => {
  return 'npm';
};
/**
 * Usage: `const path = getPackageRoot(__dirname)`
 */


exports.getNodePackageManager = getNodePackageManager;

const getPackageRoot = cwd => _pkgDir.default.sync(cwd);

exports.getPackageRoot = getPackageRoot;
